packages/firestore/src/remote/rest_connection.ts

```diff
@@ -85,7 +89,12 @@ export abstract class RestConnection implements Connection {
     const url = this.makeUrl(rpcName, path);
     logDebug(LOG_TAG, `Sending RPC '${rpcName}' ${streamId}:`, url, req);
 
-    const headers = {};
+    const headers: StringMap = {
+      'x-goog-request-params':
+        this.databaseId.database === DEFAULT_DATABASE_NAME
+          ? `project_id=${this.databaseId.projectId}`
+          : `project_id=${this.databaseId.projectId}&database_id=${this.databaseId.database}`
```

Comment:
Drive-by comment: what if these variables contain characters that require url escaping?

Reply:
Database names are restricted to regex:  /[a-z][0-9]-/ 

I am not sure about project id.

This is in a header, not url. I honestly don't know whether escaping is required in this context.

Reply:
Maybe just test it out with some of the characters from https://datatracker.ietf.org/doc/html/rfc3986.html#page-12, such as `:` , `,` , `?` , `#` , `[` , `]` , `@` `!` , `$` , `&` , `'` , `(` , `)` , `*` , `+` , `,` , `;` , and `=`. There may be a "sql injection" style vulnerability if quoting is required but not being done.

Reply:
Your proding made me dig deeper. The standard for `x-goog-request-params' says to URL encode. 
https://google.aip.dev/client-libraries/4222

I have added `google-cloud-resource-prefix` as well for backward compatibility. But it should be removed in a future client, once the current Firestore emulator is no longer in use. I am not concerned with future proofing this header, since it should be removed later. But with the current allowed characters in database names, doing URL encoding should be harmless and I will do it here as well.

The HTTP standard doesn't use URL encoding on header values. It is a completely different beast.
https://en.wikipedia.org/wiki/List_of_HTTP_header_fields#General_format

After some research, I would argue that applying URL encoding could be harmful. For example:

`Accept-Datetime: Thu, 31 May 2007 20:35:00 GMT`

Would be mangled if sent as:

`Accept-Datetime: Thu%2C%2031%20May%202007%2020%3A35%3A00%20GMT`

So I think every header we deal with needs to be understood before applying any encoding.

Reply:
I 100% agree that we shouldn't urlencode _every_ header. I didn't mean to imply that we should.

But `x-goog-request-params`, as you found, expects the %-style escaping. So it's definitely a good idea to urlencode it. I don't know about the `google-cloud-resource-prefix` header though.